File: tools/add_chapter_numbers.py
#!/usr/bin/env python3
"""Add chapterNum field to chapters in subject.js, numbered sequentially within each form."""

import re

subject_js_path = "../js/subject.js"

with open(subject_js_path, "r", encoding="utf-8") as f:
    lines = f.readlines()

output = []
form_counters = {}
i = 0

while i < len(lines):
    line = lines[i]

    # Detect form field: form:"T2", form:"T3", or form:"T4/T5" etc.
    form_match = re.search(r'form:\s*["\']([T]\d(?:/[T]\d)?)["\']', line)
    if form_match:
        form = form_match.group(1)
        # Initialize counter for this form if needed
        if form not in form_counters:
            form_counters[form] = 1

        output.append(line)

        # Check if next meaningful line has chapterNum already
        j = i + 1
        found_chapter_num = False
        while j < len(lines) and j < i + 10:
            if re.search(r'chapterNum\s*:', lines[j]):
                found_chapter_num = True
                break
            if re.search(r'chapter\s*:|resourceUrl\s*:|status\s*:', lines[j]):
                break
            j += 1

        # If no chapterNum, add it on next line
        if not found_chapter_num:
            # Insert chapterNum line after form
            indent = "            "  # Match existing indentation
            output.append(f'{indent}chapterNum: {form_counters[form]},\n')
            form_counters[form] += 1
        else:
            form_counters[form] += 1
    else:
        output.append(line)

    i += 1

with open(subject_js_path, "w", encoding="utf-8") as f:
    f.writelines(output)

print("Added chapterNum to all chapters")
print(f"Form counters: {form_counters}")

File: tools/add-chapter-numbers.js
#!/usr/bin/env node
/**
 * Add chapterNum field to chapters in subject.js
 * Numbers each chapter sequentially within its form (T2, T3, T4, T5)
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const subjectJsPath = path.join(__dirname, '../js/subject.js');

let content = fs.readFileSync(subjectJsPath, 'utf-8');
let addedCount = 0;

// Process each chapter object that doesn't already have chapterNum
// Pattern: { form: "Tx", ... } and add chapterNum after form
content = content.replace(
    /(\{\s*form:\s*"(T\d)")([^}]*?)(?=,\s*(?:chapter|resourceUrl|chinese|status))/g,
    (match, opening, form, rest) => {
        // Only add if not already present
        if (rest.includes('chapterNum')) {
            return match;
        }
        addedCount++;
        return `${opening}",\n            chapterNum: 0${rest}`;
    }
);

// Now number them sequentially within each form
const lines = content.split('\n');
const formCounters = new Map();

for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const chapterNumMatch = line.match(/chapterNum:\s*0([,\s])/);
    const formMatch = lines.slice(Math.max(0, i - 5), i)
        .join('\n')
        .match(/form:\s*"(T\d+)"/);

    if (chapterNumMatch && formMatch) {
        const form = formMatch[1];
        if (!formCounters.has(form)) {
            formCounters.set(form, 1);
        } else {
            formCounters.set(form, formCounters.get(form) + 1);
        }

        const num = formCounters.get(form);
        lines[i] = line.replace(/chapterNum:\s*0/, `chapterNum: ${num}`);
    }
}

fs.writeFileSync(subjectJsPath, lines.join('\n'), 'utf-8');
console.log(`✓ Added chapterNum to ${addedCount} chapters`);
